src/d3/hapticplot.d3.spec.ts

```diff
@@ -140,6 +137,131 @@ describe('VR Haptic Plot', () => {
     expect(getIntervalReset(controller)).toEqual(100);
   });
 
+  it('kdtree for an empty set of data points is null', () => {
+    const dataArray = [];
+    hapticplot.init(scene, dataArray);
+    expect(get3DTree(scene, shape)).toEqual(null);
+  });
+
+  it('builds a correct k-d tree for a scene with a single datapoints', () => {
+    const dataArray = [10];
+    hapticplot.init(scene, dataArray);
+    expect(get3DTree(scene, shape)).toEqual({
+      position: new Vector3(0, 1.7, -0.35),
+      left: null,
+      right: null });
+  });
+
+  it('builds a correct k-d tree of 4 datapoints', () => {
+    const dataArray = [10, 15, 20, 25];
+    hapticplot.init(scene, dataArray);
+    expect(get3DTree(scene, shape)).toEqual({
+      position: new Vector3(0.35, 1.56, -0.35),
+      left: {
+        position: new Vector3(0.175, 1.42, -0.35),
+        left: {
+          position: new Vector3(0, 1.28, -0.35),
+          left: null,
+          right: null,
+        },
+        right: null,
+      },
+      right: {
+        position: new Vector3( 0.5249999999999999, 1.7, -0.35),
+        left: null,
+        right: null,
+      },
+    });
+  });
+
+  it('builds a correct k-d tree of 4 datapoints with increasing and deceasing values', () => {
+    const dataArray = [10, 15, 18, 13, 9, 17];
+    hapticplot.init(scene, dataArray);
+    expect(get3DTree(scene, shape)).toEqual({
+      position: new Vector3(0.35, 1.5055555555555555, -0.35),
+      left: {
+        position: new Vector3(0.11666666666666665, 1.5833333333333335, -0.35),
+        left: {
+          position: new Vector3(0, 1.3888888888888888, -0.35),
+          left: null,
+          right: null,
+        },
+        right: {
+             position: new Vector3(0.2333333333333333, 1.7, -0.35),
+             left: null,
+             right: null,
+        }
+      },
+      right: {
+        position: new Vector3(0.5833333333333333, 1.661111111111111, -0.35),
+        left: {
+          position: new Vector3(0.4666666666666666, 1.35, -0.35),
+          left: null,
+          right: null,
+        },
+        right: null,
+      }
+    });
+  });
+
+  it('builds a correct k-d tree for a scene with multiple datapoints with the same value', () => {
+    const dataArray = [5, 5, 5, 5];
+    hapticplot.init(scene, dataArray);
+    expect(get3DTree(scene, shape)).toEqual({
+      position: new Vector3(0.35, 1.7, -0.35),
+      left: {
+        position: new Vector3(0.175, 1.7, -0.35),
+        left: {
+          position: new Vector3(0, 1.7, -0.35),
+          left: null,
+          right: null,
+        },
+        right: null,
+      },
+      right: {
+        position: new Vector3(0.5249999999999999, 1.7, -0.35),
+        left: null,
+        right: null,
+      },
+    });
+  });
```

Comment:
All the stuff about building and probing the kd-tree should be in a separate test spec file.  Then if you need to move the kd stuff elsewhere, it's obvious what needs to go.  It also makes for better decoupling.